controller/volume_attachment_controller.go

```diff
@@ -0,0 +1,413 @@
+			return
+		}
+		_, err = vac.ds.UpdateLHVolumeAttachmetStatus(va)
+		return
+	}()
+
+	vol, err = vac.handleVolumeDetachment(va, vol)
+	if err != nil {
+		return err
+	}
+
+	vol, err = vac.handleVolumeAttachment(va, vol)
+	if err != nil {
+		return err
+	}
+
+	return vac.handleVAStatusUpdate(va, vol)
+}
+
+func (vac *VolumeAttachmentController) handleVolumeDetachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) (*longhorn.Volume, error) {
+	// Volume is already trying to detach
+	if vol.Spec.NodeID == "" {
+		return vol, nil
+	}
+
+	for _, attachment := range va.Spec.Attachments {
+		// Found one attachment that is still requesting volume to attach to the current node
+		if attachment.NodeID == vol.Spec.NodeID && verifyAttachmentParameters(attachment, vol) {
+			return vol, nil
+		}
+	}
+
+	// There is no attachment that request the current vol.Spec.NodeID.
+	// Therefore, set desire state of volume to empty
+	vol.Spec.NodeID = ""
+	// reset the attachment parameter
+	setAttachmentParameter(&longhorn.Attachment{}, vol)
+	return vac.ds.UpdateVolume(vol)
+}
+
+func (vac *VolumeAttachmentController) handleVolumeAttachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) (*longhorn.Volume, error) {
+	// Wait for volume to be ready to be fully detached
+	if vol.Spec.NodeID != "" ||
+		vol.Spec.MigrationNodeID != "" ||
+		vol.Status.PendingNodeID != "" ||
+		vol.Status.State != longhorn.VolumeStateDetached {
+		return vol, nil
+	}
+
+	attachment := selectAttachmentToAttach(va)
+	if attachment == nil {
+		return vol, nil
+	}
+
+	vol.Spec.NodeID = attachment.NodeID
+	setAttachmentParameter(attachment, vol)
+	return vac.ds.UpdateVolume(vol)
+}
+
+func selectAttachmentToAttach(va *longhorn.VolumeAttachment) *longhorn.Attachment {
+	if len(va.Spec.Attachments) == 0 {
+		return nil
+	}
+
+	highPriorityAttachments := []*longhorn.Attachment{}
+	maxAttacherPriorityLevel := 0
+	for _, attachment := range va.Spec.Attachments {
+		priorityLevel := longhorn.GetAttacherPriorityLevel(attachment.Type)
+		if priorityLevel > maxAttacherPriorityLevel {
+			maxAttacherPriorityLevel = priorityLevel
+		}
+	}
+
+	for _, attachment := range va.Spec.Attachments {
+		priorityLevel := longhorn.GetAttacherPriorityLevel(attachment.Type)
+		if priorityLevel == maxAttacherPriorityLevel {
+			highPriorityAttachments = append(highPriorityAttachments, attachment)
+		}
+	}
+
+	// TODO: sort by time
+
+	// sort by name
+	shortestNameAttachment := highPriorityAttachments[0]
+	for _, attachment := range highPriorityAttachments {
+		if attachment.ID < shortestNameAttachment.ID {
+			shortestNameAttachment = attachment
+		}
+	}
+
+	return shortestNameAttachment
+}
+
+func (vac *VolumeAttachmentController) handleVAStatusUpdate(va *longhorn.VolumeAttachment, vol *longhorn.Volume) error {
+	// initialize the va.Status.Attachments map if needed
+	if va.Status.Attachments == nil {
+		va.Status.Attachments = make(map[string]*longhorn.Attachment)
+	}
```

Comment:
Can we do this in the mutating webhook?

Reply:
I agree. But just to safeguard against the nill map insertion, should we also initialize it here?